src/rgw/rgw_sal_motr.h

```diff
@@ -40,6 +44,215 @@ class MotrStore;
 #define RGW_MOTR_BUCKET_HD_IDX_NAME   "motr.rgw.bucket.headers"
 //#define RGW_MOTR_BUCKET_ACL_IDX_NAME  "motr.rgw.bucket.acls"
 
+// Notes for Motr FDMI:
+// (1) Currently FDMI can only trigger events for index operation, not object
+//     operation.
+// (2) When an index op event is generated, the FDMI filter picks up the events
+//     matching the pre-defined rules. The FDMI event contains the key/value of
+//     the index operation. Currently, fdmi can only do sub string matching on
+//     the value part, not key part, and can't specify which index to watch,
+//     the filter rule is applied to all indices.
+//     this 
+// (3) The FDMI event is delivered to an FDMI application which defines a
+//     callback function to process FDMI events.
+//
+// Implementation of object watch-notify using Motr FDMI. 
+// (1) An FDMI filter is added (in Motr configuration yaml file) specifying
+//     a matching sub-string (marker).
+// (2) A notifier creates a set of indices which are used to trigger FDMI
+//     events and to pass messages.
+// (3) For an object, hash(obj_name) is used to pick which index to insert
+//     a new key/value record,
+//     key = unique_fid(obj_name),
+//     value = marker (for fdmi filter) + notification msg.
+//     If multiple rgw instances are trying to write the same objects at the same
+//     time, we assume that Motr index is updated atomically.
+// (4) An FDMI event with a special marker is generated and
+//     picked up by the filter and delivered to our fdmi application, watcher.
+//     The watcher defines the callback function processing the FDMI events.
+// (5) The embedded notification message is then decoded to get object name
+//     and index opcode for further action.
+//
+// How metadata cache use watch-notify:
+// (1) MGW uses an index to store object metatadata.
+// (2) As metadata has its own index, instead of creating a set of new
+//     notification indices above, metadata cache makes use of the metadata
+//     index to pass notification.
+//     When an object's metadata is updated, the corresponding key/value pair
+//     in the metadata index is updated. The value is attached with a
+//     notification message containing the FDMI filter's matching sub-string.
+// (3) When the watcher receives notification, the decoded info are sent to
+//     cache layer.
+// (4) Update cache item accordingly and release notification(fdmi record).
+// (5) Note: if the size of metadata is big, the FDMI event size created will
+//     be big too (key + notification + metadata), the overhead to transfer
+//     the FDMI event is big, how will this impact the performance and
+//     scalability of FDMI?
```

Comment:
Good point, it should be kept in mind.

In general, I think, we should avoid big metadata records. If there is something big - it probably would not be needed to be read very often, so we may offload it into a separate index, as @nikitadanilov pointed out at the last Meet Architect meeting.